gap/attr.gi

```diff
@@ -17,85 +17,152 @@ InstallMethod(OutNeighbours, "for a digraph by out-neighbours",
 # The next method is (yet another) DFS as described in
 # http://www.eecs.wsu.edu/~holder/courses/CptS223/spr08/slides/graphapps.pdf
 
-InstallMethod(ArticulationPoints, "for a digraph by out-neighbours",
-[IsDigraphByOutNeighboursRep],
+BindGlobal("DIGRAPHS_ArticulationPointsBridgesStrongOrientation",
 function(D)
-  local copy, nbs, counter, visited, num, low, parent, points, points_seen,
-        stack, depth, v, w, i;
-  if (HasIsConnectedDigraph(D) and not IsConnectedDigraph(D))
-      or DigraphNrVertices(D) <= 1 then
-    return [];
+  local N, copy, articulation_points, bridges, orientation, nbs, counter, pre,
+        low, nr_children, stack, u, v, i, w, connected;
+
+  N := DigraphNrVertices(D);
+
+  if HasIsConnectedDigraph(D) and not IsConnectedDigraph(D) then
+    # not connected, no articulation points, no bridges, no strong orientation
+    return [false, [], [], fail];
+  elif N < 2 then
+    # connected, no articulation points (removing 0 or 1 nodes does not make
+    # the graph disconnected), no bridges, strong orientation (since
+    # the digraph with 0 nodes is strongly connected).
+    return [true, [], [], D];
   elif not IsSymmetricDigraph(D) then
     copy := DigraphSymmetricClosure(DigraphMutableCopyIfMutable(D));
+    MakeImmutable(copy);
   else
     copy := D;
   fi;
+
+  # outputs
+  articulation_points := [];
+  bridges             := [];
+  orientation         := List([1 .. N], x -> BlistList([1 .. N], []));
+
+  # Get out-neighbours once, to avoid repeated copying for mutable digraphs.
   nbs := OutNeighbours(copy);
 
-  counter     := 0;
-  visited     := BlistList([1 .. DigraphNrVertices(copy)], []);
-  num         := [];
-  low         := [];
-  parent      := [1];
-  points      := [];
-  points_seen := BlistList([1 .. DigraphNrVertices(copy)], []);
-  stack       := [[1, 0]];
-  depth       := 1;
-
-  while depth > 1 or not visited[1] do
-    v := stack[depth][1];
-    if visited[v] then
-      depth := depth - 1;
-      v     := stack[depth][1];
-      w     := nbs[v][stack[depth][2]];
-      if v <> 1 and low[w] >= num[v] and not points_seen[v] then
-        points_seen[v] := true;
-        Add(points, v);
+  # number of nodes encountered in the search so far
+  counter := 0;
+
+  # the order in which the nodes are visited, -1 indicates "not yet visited".
+  pre := ListWithIdenticalEntries(N, -1);
```

Comment:
I was following the reference given on Line 18, and it refers to this ordering as 'Num'. Since broadly the variable names are in keeping with the reference, might it be better to call this list `num` instead of `pre` for ease of comparison?

Reply:
Maybe I should update the reference, which only explains how to compute the articulation points, there's another reference for computing bridges (which involves adding 3 lines to the code for articulation points), there `pre` is used, I'll add a reference to this too in a comment.